P2 signup: P2 Complete Profile

lamosty opened the pull request:
Addresses 4741-gh-automattic/p2.

In this PR, we add a new step to the P2 signup flow:

<img width="1498" alt="Screenshot on 2022-03-18 at 18-04-35" src="https://user-images.githubusercontent.com/4988512/159049554-6600fe04-5281-4aa5-9646-1199d03c6b44.png">

It allows uploading or modifying the user's Gravatar and changing their full name (`display_name`).

## Testing instructions

With an incognito window, navigate to `/start/p2`. Using a 10 minute email address, create a new user and verify the email. In the first tab, reload the signup flow (as the email confirmation link will take you to wordpress.com/production). After clicking on Continue, you should arrive to the Complete Profile step.

Try to upload a Gravatar and change your full name. The name must be longer than 3 characters and the Gravatar must be an image.

lamosty committed:
P2 signup: init P2 Complete Profile

lamosty committed:
Use class-based approach

lamosty committed:
Add skip btn and style it

lamosty committed:
Add Timezone dropdown and style it

lamosty committed:
Implement skip btn

lamosty committed:
Guess user timezone and set as initial value

lamosty committed:
Save user display name

lamosty committed:
Modify EditGravatar to accept additional html to open file picker

lamosty committed:
Remove timezone, add validation to full name

lamosty commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -234,6 +234,11 @@ export class EditGravatar extends Component {
 							}
 						) }
 					</InfoPopover>
+					{ additionalUploadHtml && (
```
Had to modify the original `EditGravatar` component to accomplish this.

tyxla replied:
I wonder if we could have declared those as child components and just used the `children` prop here.

lamosty replied:
Yeah, not sure. Should it have children, though? I'd say this is more of a special case as what would be the children of `EditGravatar` in the logical sense?

I'll defer to your call.

tyxla replied:
It's just at the end of the component and having it as `children` made sense to me. It also makes the component more generic and less specific, and also, you're using it as a similar syntax as children already, just with a different prop name 😉 . Happy to use `children` for that.

lamosty replied:
Great, will do. 👍

lamosty committed:
Move the step to proper place in the flow

lezama commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+
+					<div className="p2-complete-profile__skip-wrapper">
+						<span>
+							No time? No problem! You can{ ' ' }
```
missing translation?

tyxla replied:
This. Also the string seems unfinished.

lezama commented:
works great, let's ship this!

@tyxla could you give this a quick 👀 ? 🙏

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+import { Button } from '@wordpress/components';
+import { localize } from 'i18n-calypso';
+import { isEmpty } from 'lodash';
```
By all means, let's avoid introducing new Lodash instances. We've been refactoring away from it for a while now, see p4TIVU-9Bf-p2

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+		if ( props?.step?.form ) {
+			initialState = props.step.form;
+
+			if ( ! isEmpty( props.step.errors ) ) {
```
Seems like `props.step.errors` will be an array, so `.length` should work instead of using Lodash's `isEmpty()`

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+	validate = ( fields, onComplete ) => {
+		const messages = {};
+
+		if ( isEmpty( fields.fullName ) || fields.fullName.length < 3 ) {
```
Seems like `fields.fullName` will be an array, so `.length` should work instead of using Lodash's `isEmpty()`

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+
+		let initialState;
+
+		if ( props?.step?.form ) {
```
Is there really a possibility for `props` to be nullish?

What about the `step` prop?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+	validate = ( fields, onComplete ) => {
+		const messages = {};
+
+		if ( isEmpty( fields.fullName ) || fields.fullName.length < 3 ) {
```
Noting that the artificial limitations on names could be discriminatory. I'm not convinced that one can't have a 2-character full name nowadays.

lamosty replied:
Changed to 2 chars.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+ * Constants
+ */
+const VALIDATION_DELAY_AFTER_FIELD_CHANGES = 1500;
+const ERROR_CODE_MISSING_FULL_NAME = 123; // Random number, we don't need it.
```
If we don't need the number, can we just omit setting a special array key?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	componentWillUnmount() {
+		this.save();
```
Won't this cause memory leaks or weird side effects due to the fact that requests are performed on behalf of a component that has already been unmounted?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+ */
+const VALIDATION_DELAY_AFTER_FIELD_CHANGES = 1500;
+const ERROR_CODE_MISSING_FULL_NAME = 123; // Random number, we don't need it.
+const ERROR_CODE_FROM_LOCAL_STORAGE = 7331; // Random number, we don't need it.
```
Same as for the above constant.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+				this.setState( { form: state } );
+			},
+			onError: this.handleFormControllerError,
+			debounceWait: VALIDATION_DELAY_AFTER_FIELD_CHANGES,
```
Any good reason not to declare the debounce timeout inline?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+			onError: this.handleFormControllerError,
+			debounceWait: VALIDATION_DELAY_AFTER_FIELD_CHANGES,
+			hideFieldErrorsOnChange: true,
+			initialState: initialState,
```
Can be shortened to:

```suggestion
			initialState,
```

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+			form: this.formStateController.getInitialState(),
+			isSubmitting: false,
+			suggestedSubdomains: [],
+			lastInvalidSite: '',
```
This state property doesn't seem to be used.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+		this.state = {
+			form: this.formStateController.getInitialState(),
+			isSubmitting: false,
+			suggestedSubdomains: [],
```
This state property doesn't seem to be used.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+			if ( ! isEmpty( props.step.errors ) ) {
+				const errorMessage = props.step.errors[ 0 ].message;
+
+				this.logValidationErrorToLogstash( ERROR_CODE_FROM_LOCAL_STORAGE, errorMessage );
```
`this.logValidationErrorToLogstash()` is undefined 🤔  Did we mean to copy it from the site step? Do we actually need this logging? Can we remove it if not?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+
+				this.logValidationErrorToLogstash( ERROR_CODE_FROM_LOCAL_STORAGE, errorMessage );
+
+				initialState = formState.setFieldErrors(
```
I find it a bit confusing that `initialState` is being mutated and has so much logic in the first place. Usually, initial state is static and if it changes during the component's lifecycle, it's practically no longer "initial".

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+
+		return (
+			<>
+				<ValidationFieldset
```
I find the entire validation logic here extremely complex just for the sake of validating a single text field. I bet we can write the validation in a much simpler way and much less code without using `formState`. Have you tried that? I'd very much recommend avoiding it for such simple use cases.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+		const fieldDisabled = this.state.isSubmitting;
+
+		return (
+			<>
```
Since `<ValidationFieldset />` is the sole child component of this fragment, we can remove the fragment altogether.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+					</FormLabel>
+					<FormTextInput
+						id="full-name-input"
+						autoFocus={ true } // eslint-disable-line jsx-a11y/no-autofocus
```
Auto-focusing the fields is not recommended for a good reason. Are we certain we really need it here?

In any case, `={ true }` is redundant:

```suggestion
						autoFocus // eslint-disable-line jsx-a11y/no-autofocus
```

Although I'd vote for removing this prop completely.

lamosty replied:
Will do.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+				initialState = formState.setFieldErrors(
+					formState.setFieldsValidating( initialState ),
+					{
+						site: {
```
This seems wrong - I don't see a `site` field in this component.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+
+		this.formStateController = new formState.Controller( {
+			fieldNames: [ 'fullName' ],
+			sanitizerFunction: this.sanitize,
```
I don't see a `sanitize` method declared in this component. Perhaps we can just ditch this line?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+		const messages = {};
+
+		if ( isEmpty( fields.fullName ) || fields.fullName.length < 3 ) {
+			messages.fullName = {
```
Can we avoid mutating objects like that? Mutability can cause terrible bugs.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+
+			// API calls
+
+			this.props.saveUserSettings( { display_name: fullName } );
```
What happens if this request fails? Are we handling it in any way?

lamosty replied:
Right now, the Calypso's data layer handles it by showing a "There was a problem saving your changes." notice. Not sure how rare it would be and whether we need to intercept it in this step and tell user to try again or something or just continue the flow as we do now.

tyxla replied:
IMHO, the probability for it to fail is the same as any other request. The difference is that we do handle some of those failures gracefully, while here, we'd likely fail silently (silently for the user at least).

lamosty replied:
Yep, probability is the same but I guess what I meant is that this is not an important request. Worst case user will not have their name set. The whole step is kind of additional. But yeah, I'll try to intercept the error and stop the submit in that case.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+				fullName,
+			};
+
+			this.props.submitSignupStep( stepData );
```
What happens if this request fails? Are we handling it in any way?

Also do we need to wait for it to finish?

lamosty replied:
`submitSignupStep` is the regular signup facility to continue the flow. I'm not sure if there could be a failure in submitting the step and the other steps just submit it, too. If the API request to update user full name which is called above this fails we could perhaps stop the signup process and tell the user to try again?

tyxla replied:
Sounds good, any kind of failure handling makes sense when there's a request in place that could possibly fail.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+		} );
+	};
+
+	handleFormControllerError = ( error ) => {
```
Abstractions like that are why I really recommend not using `formState`. Just for the protocol 😉

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+					</div>
+
+					<div className="p2-complete-profile__skip-wrapper">
+						<span>
```
Do we need the span? If we need it for styling, can we add a classname and style by that classname?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+							<Button
+								className="p2-complete-profile__skip-btn"
+								variant="link"
+								onClick={ () => {
```
Can we abstract that callback into a method so it doesn't get redeclared on every re-render?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,260 @@
+									this.props.goToNextStep();
+								} }
+							>
+								do this later.
```
Another part of the string that is missing translation here.

tyxla reviewed (commented):
> @tyxla could you give this a quick 👀 ? 🙏

Sure thing @lezama 👍

lamosty committed:
Initial component refactor to stateless

lamosty committed:
Add a bit more styling

lamosty committed:
Properly add the skip btn

lamosty committed:
Add better styling override for the primary btn

lamosty committed:
Refactor EditGravatar to accept children

lamosty committed:
Move back the signup step in the flow

lamosty committed:
Remove stray css

lamosty committed:
Abstract skip btn callback

lamosty committed:
Change fullName validation

lamosty committed:
Remove auto focus

lamosty commented:
After a rebase, the primary button is broken (recent issue). We are working on a fix in another PR. Will rebase afterwards.

tyxla commented on client/blocks/edit-gravatar/index.jsx:
```diff
@@ -231,6 +231,11 @@ export class EditGravatar extends Component {
 						) }
 					</InfoPopover>
+					{ children && (
+						<FilePicker accept="image/*" onPick={ this.onReceiveFile }>
```
Hm, so we do need `this.onReceiveFile` out here. Guess this wasn't such a good use for `children` after all 😭 Mind reverting this change? I owe you a beer.

lamosty replied:
Hehe, sure. Done. 😅

tyxla commented:
> After a rebase, the primary button is broken (recent issue). We are working on a fix in another PR. Will rebase afterwards.

I see much (if not all) of the feedback has been addressed. Let me know when this one is ready for a review and I'll take another look. Thanks!

lamosty committed:
Revert "Refactor EditGravatar to accept children"

This reverts commit fa6a80d0498e499af88244faa56a9fe62e06c8da.

lamosty committed:
Submit step only if no errors when saving user settings

lamosty commented:
@tyxla I added a simple error handling in [f6b4d1d](https://github.com/Automattic/wp-calypso/pull/61914/commits/f6b4d1d68670acbe3c47c9b0ced8768022e531de) but not sure if it's the way to go. Could you give it another pass, please? I think I've addressed all the other points.

annemirasol commented:
Good job, Rasto!

I noticed that it makes me go through the Profile step again, even though I have already submitted an avatar and name in a previous signup flow.

1. Go through signup flow and complete your profile. Optionally, create a workspace.
2. Go to `calypso.localhost:3000/start/p2` again, and notice that you get the profile page again.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,163 @@
+				</div>
+				<div className="p2-complete-profile__skip-wrapper">
+					{ createInterpolateElement(
+						__( 'No time? No problem! You can <Button>do this later</Button>.' ),
```
I find it very odd that we use both `i18n-calypso` and `@wordpress/i18n` for translations. Can we use just one of them in a single component?

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,163 @@
+			}
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [ isSubmitting, updatingUserSettings, userSettingsRequestFailed ] );
```
Any particular reason why we're not listing all dependencies of this hook here?

lamosty replied:
I was thinking to include only those which could actually change and be important to react to their change. The others are just functions and stuff. I'll include them all.

tyxla commented on client/signup/steps/p2-complete-profile/index.jsx:
```diff
@@ -0,0 +1,163 @@
+
+		if ( formFullName.length < 2 ) {
+			setFormErrors( {
+				fullName: translate( 'Please enter your full name (3 characters or more).' ),
```
Seems like we require at least 2 characters but the error says at least 3?

tyxla commented on client/signup/steps/p2-complete-profile/style.scss:
```diff
@@ -0,0 +1,77 @@
+	}
+}
+
+.p2-complete-profile__avatar-wrapper {
```
We already declare styles for this particular selector at the top of the file. Should we merge both style blocks?

tyxla reviewed (commented):
Haven't tested, just adding a few more suggestions on the code before we go ahead to merge this.

lamosty committed:
Skip step if user has display name already set

lamosty commented:
@annemirasol I've added the logic to skip the step if user has already filled in their display name in [a15467e](https://github.com/Automattic/wp-calypso/pull/61914/commits/a15467e750f31fde94649b379dc95649acc971da). I wanted to add the same logic for the Gravatar but at the moment I don't know how to distinguish between the default Gravatar url and a custom one. The default Gravatar url is different for every user.

lamosty committed:
Address feedback

lamosty commented:
Thanks for the very detailed review @tyxla! Points addressed.

lezama reviewed (approved):
let's 🚢 this! Thanks @tyxla for your 👀 once more!

annemirasol commented:
> I've added the logic to skip the step if user has already filled in their display name in [a15467e](https://github.com/Automattic/wp-calypso/pull/61914/commits/a15467e750f31fde94649b379dc95649acc971da). I wanted to add the same logic for the Gravatar but at the moment I don't know how to distinguish between the default Gravatar url and a custom one. The default Gravatar url is different for every user.

I think this is okay and may be best, actually. I personally don't want to be bugged more than once about uploading a personal avatar. The name is a different matter, since we tend to see a lot of usernames that are unrelated to the person's name.

Nit: maybe `excludeStepIfProfileComplete`? `details` just seem general.

tyxla reviewed (approved):
Haven't tested, but the code's looking great!

lamosty committed:
Merge branch 'trunk' into p2-signup/add-complete-your-profile-step

lamosty committed:
Remove changes to composer.json

lamosty committed:
Refactor to excludeStepIfProfileComplete

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7209197

Thank you @lamosty for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7209197).
